R/utils-ui.R

```diff
@@ -271,3 +271,89 @@ compute_n_show <- function(n, n_show_nominal = 5, n_fudge = 2) {
     n
   }
 }
+
+# menu(), but based on readline() + cli and mockable ---------------------------
+# https://github.com/r-lib/cli/issues/228
+# https://github.com/rstudio/rsconnect/blob/main/R/utils-cli.R
+
+cli_menu <- function(header,
+                     prompt,
+                     choices,
+                     not_interactive = choices,
+                     exit = integer(),
+                     .envir = caller_env(),
+                     error_call = caller_env()) {
+  if (!is_interactive()) {
+    cli::cli_abort(
+      c(header, not_interactive),
+      .envir = .envir,
+      call = error_call
+    )
+  }
+
+  choices <- paste0(cli::style_bold(seq_along(choices)), ": ", choices)
+  cli::cli_inform(
+    c(header, prompt, choices),
+    .envir = .envir
+  )
+
+  # guard against invalid mocked input
+  local_input <- getOption("cli_input", character())
+
+  repeat {
+    selected <- cli_readline("Selection: ")
+    if (selected %in% c("0", seq_along(choices))) {
+      break
+    }
+    if (length(local_input) > 0) {
+      cli::cli_abort(
+        c(x = "Internal error: mocked input is invalid."),
+        .envir = .envir,
+        call = error_call
+      )
+    }
```

Comment:
The second part of "avoiding an infinite loop".

Reply:
Can we do that all in one place? Or extract it out into a helper?

Reply:
OK I consolidated things, just to make it a bit neater, at least.

Reply:
Why does it error if it’s length greater than zero? The original intent was to be able to supply multiple values that were used in turn.

Reply:
Over in rsconnect, if I add this test, it causes the test suite to just hang:

```
test_that("cli_menu() works with multiple inputs", {
  simulate_user_input(3)

  expect_snapshot(
    cli_menu(
      "Let's talk",
      "Are you OK?",
      choices = c("Yes", "No")
    )
  )
})
```
<img width="436" alt="Screenshot 2023-04-04 at 9 24 46 PM" src="https://user-images.githubusercontent.com/599454/229980725-59c1e22b-bb5b-49f5-84e8-9d51689ead4a.png">

I guess I haven't found an organic need to supply multiple values. But I _have_ managed to supply an invalid one. So it seems nice for development purposes to handle that case.

I can try to think about how to preserve both.

Reply:
Are you convinced that multiple mocked inputs is important?

It feels like we could just make multiple calls to `simulate_user_input()` (rsconnect) or `local_user_input()` (here in gargle) in such a test and let the functions be simpler.

Reply:
I realize I can pursue the simplification that's most natural here. I'm mostly talking about what we might eventually want in cli or in some standalone file for general use.